### PR TITLE
[es8311] Add es8311 dac component

### PR DESCRIPTION
**Pull request by kahrendt**
# What does this implement/fix?

Adds support for the ES8311 codec chip as an ``audio_dac`` component. The S3 Box 3 uses this for handling audio output. It supports volume and mute control in hardware.

I built this from the unmerged #4861, and so I listed @kroimon as a codeowner. If you would prefer not to be included, please let me know. I adapted that code to the ``audio_dac`` platform and exposed more yaml configuration options.

Note that this only handles audio out on the S3 box, not audio in. So it only removes the requirement for the ``esp_adf`` component's speaker. I haven't tested pairing this component with the ``esp_adf`` microphone in a voice assistant, so that may not work!

Requires the already merged #7664 to handle volume control with the speaker component.

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Code quality improvements to existing code or addition of tests
- [ ] Other

**Related issue or feature (if applicable):**

- partially implements esphome/feature-requests#2239

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):**

- esphome/esphome-docs#4400

## Test Environment

- [ ] ESP32
- [x] ESP32 IDF
- [ ] ESP8266
- [ ] RP2040
- [ ] BK72xx
- [ ] RTL87xx

## Example entry for `config.yaml`:

```yaml
# Example config.yaml

# Sets up the S3 Box 3's speaker

external_components:
  - source: github://pr#7672
    components:
      - audio
      - speaker
      - i2s_audio
    refresh: 0s

audio_dac:
  - platform: es8311
    id: es8311_dac
    bits_per_sample: 16bit
    sample_rate: 16000

i2s_audio:
  - id: i2s_output
    i2s_lrclk_pin: GPIO45
    i2s_bclk_pin: GPIO17
    i2s_mclk_pin: GPIO23

speaker:
  - platform: i2s_audio
    i2s_audio_id: i2s_output
    id: speaker_id
    i2s_dout_pin: GPIO15
    dac_type: external
    sample_rate: 16000
    bits_per_sample: 16bit
    channel: stereo
    audio_dac: es8311_dac

switch:
  - platform: gpio
    name: "Speaker Enable"
    pin: GPIO46
    restore_mode: RESTORE_DEFAULT_ON

number:
  - platform: template
    name: "Speaker Volume"
    optimistic: true
    min_value: 0.0
    max_value: 1.0
    step: 0.05
    on_value:
      speaker.volume_set: !lambda "return x;"

```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

File: esphome/components/es8311/es8311.cpp [hunk 1]
```diff
@@ -0,0 +1,239 @@
...
+  if (this->is_failed()) {
+    ESP_LOGCONFIG(TAG, "  Failed to initialize!");
+    return;
+  }
+
+#ifdef ESPHOME_LOG_HAS_VERBOSE
+  ESP_LOGV(TAG, "  Register Values:");
+  for (uint8_t reg = 0; reg <= 0x45; reg++) {
+    uint8_t value;
+    if (!ES8311_READ_BYTE(reg, &value)) {
```
Comment (jesserockz) [inline esphome/components/es8311/es8311.cpp]: We don't allow reading of data inside `dump_config`.

File: esphome/components/es8311/es8311.cpp [hunk 2]
```diff
@@ -0,0 +1,239 @@
...
+    this->mark_failed(); \
+    return; \
+  }
+// Return false; use outside of setup
+#define ES8311_ERROR_CHECK(func) \
+  if (!(func)) { \
+    return false; \
+  }
+#define ES8311_READ_BYTE(reg, value) this->read_byte(reg, value)
+#define ES8311_WRITE_BYTE(reg, value) this->write_byte(reg, value)
```
Comment (jesserockz) [inline esphome/components/es8311/es8311.cpp]: Are these really needed?

**Review by jesserockz — CHANGES_REQUESTED** [review]

File: esphome/components/es8311/es8311.cpp (hunk 1 above)
Comment (kahrendt) [reply esphome/components/es8311/es8311.cpp]: I've dropped it entirely, it shouldn't be necessary.

File: esphome/components/es8311/es8311.cpp (hunk 2 above)
Comment (kahrendt) [reply esphome/components/es8311/es8311.cpp]: Nope! I'm calling them directly now.

**Review by jesserockz — APPROVED** [review]

**Pull request merged**